File: tasks/one_task/get_scope/get_scope.go
package main

import (
 "fmt"
 "math"
 "math/rand"
 "time"
)

const TIMESTAMPS_COUNT = 50000

const PROBABILITY_SCORE_CHANGED = 0.0001

const PROBABILITY_HOME_SCORE = 0.45

const OFFSET_MAX_STEP = 3

type Score struct {
 Home int
 Away int
}

type ScoreStamp struct {
 Offset int
 Score  Score
}

func main() {
 var stamps = fillScores()

 for _, stamp := range *stamps {
  fmt.Printf("%v: %v -- %v\n", stamp.Offset, stamp.Score.Home, stamp.Score.Away)
 }

}

func generateStamp(previousValue ScoreStamp) ScoreStamp {
 random := rand.New(rand.NewSource(time.Now().UnixNano()))

 scoreChanged := random.Float32() > 1-PROBABILITY_SCORE_CHANGED
 homeScoreChange := 0
 if scoreChanged && random.Float32() > 1-PROBABILITY_HOME_SCORE {
  homeScoreChange = 1
 }

 awayScoreChange := 0
 if scoreChanged && homeScoreChange == 0 {
  awayScoreChange = 1
 }

 offsetChange := int(math.Floor(random.Float64()*OFFSET_MAX_STEP)) + 1

 return ScoreStamp{
  Offset: previousValue.Offset + offsetChange,
  Score: Score{
   Home: previousValue.Score.Home + homeScoreChange,
   Away: previousValue.Score.Away + awayScoreChange,
  },
 }
}

func fillScores() *[]ScoreStamp {

 scores := make([]ScoreStamp, TIMESTAMPS_COUNT)
 prevScore := ScoreStamp{
  Offset: 0,
  Score:  Score{Home: 0, Away: 0},
 }
 scores[0] = prevScore

 for i := 1; i < TIMESTAMPS_COUNT; i++ {
  scores[i] = generateStamp(prevScore)
  prevScore = scores[i]
 }

 return &scores
}

/*
 Takes list of game's stamps and time offset for which returns the scores for the home and away teams.
    Please pay attention to that for some offsets the game_stamps list may not contain scores.
*/

File: tasks/one_task/get_scope/get_scope.py
from pprint import pprint
import random
import math

TIMESTAMPS_COUNT = 50000

PROBABILITY_SCORE_CHANGED = 0.0001

PROBABILITY_HOME_SCORE = 0.45

OFFSET_MAX_STEP = 3

INITIAL_STAMP = {
    "offset": 0,
    "score": {
        "home": 0,
        "away": 0
    }
}

def main():
    game_stamps = generate_game()
    pprint(game_stamps)
    

def generate_stamp(previous_value):
    score_changed = random.random() > 1 - PROBABILITY_SCORE_CHANGED
    home_score_change = 1 if score_changed and random.random() > 1 - \
                             PROBABILITY_HOME_SCORE else 0
    away_score_change = 1 if score_changed and not home_score_change else 0
    offset_change = math.floor(random.random() * OFFSET_MAX_STEP) + 1

    return {
        "offset": previous_value["offset"] + offset_change,
        "score": {
            "home": previous_value["score"]["home"] + home_score_change,
            "away": previous_value["score"]["away"] + away_score_change
        }
    }


def generate_game():
    stamps = [INITIAL_STAMP, ]
    current_stamp = INITIAL_STAMP
    for _ in range(TIMESTAMPS_COUNT):
        current_stamp = generate_stamp(current_stamp)
        stamps.append(current_stamp)

    return stamps


def get_score(game_stamps, offset):
    """
        Takes list of game's stamps and time offset for which returns the scores for t
        he home and away teams.
        Please pay attention to that for some offsets the game_stamps list 
        may not contain scores.
    """
    if game_stamps[-1].get("offset") < offset:
        score = game_stamps[-1].get("score")
        home = score.get("home") if score else None
        away = score.get("away") if score else None
        if type(home) == int and type(away) == int:
            return home, away

    while offset >= 0:
        for stamp in game_stamps:
            if stamp.get("offset") == offset:
                score = stamp.get("score")
                try:
                    home = score.get("home") if score else None
                    away = score.get("away") if score else None
                    return home, away
                except ValueError as err:
                    print(err)
                    
        offset = offset - 1
    return 0, 0


if __name__ == "__main__":
   main()